plugins/GSdx/Renderers/SW/GSSetupPrimCodeGenerator.all.cpp

```diff
@@ -0,0 +1,572 @@
+/*
+ *	Copyright (C) 2007-2009 Gabest
+ *	http://www.gabest.org
+ *	Copyright (C) 2020 PCSX2 Dev Team
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with GNU Make; see the file COPYING.  If not, write to
+ *  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA USA.
+ *  http://www.gnu.org/copyleft/gpl.html
+ *
+ */
+
+#include "stdafx.h"
+#include "GSSetupPrimCodeGenerator.all.h"
+#include "Renderers/SW/GSScanlineEnvironment.h"
+#include "Renderers/Common/GSFunctionMap.h"
+#include "GSUtil.h"
+#include "GSVertexSW.h"
+
+using namespace Xbyak;
+
+// GCC doesn't like `offsetof(thing, arr[localVariable])`
+#if defined(__GNUC__) && !defined(__clang__)
+# undef offsetof
+# define offsetof(a, b) ((size_t)&(static_cast<a*>(0)->b))
+#endif
+
+#define _rip_local(field) ((is32 || m_rip) ? ptr[rip + (char*)&m_local.field] : ptr[_m_local + offsetof(GSScanlineLocalData, field)])
+
+#define _64_m_local _64_t0
+
+/// On AVX, does a v-prefixed separate destination operation
+/// On SSE, moves src1 into dst using movdqa, then does the operation
+#define THREEARG(operation, dst, src1, ...) \
+	do { \
+		if (hasAVX) \
+		{ \
+			v##operation(dst, src1, __VA_ARGS__); \
+		} \
+		else \
+		{ \
+			movdqa(dst, src1); \
+			operation(dst, __VA_ARGS__);\
+		} \
+	} while (0)
+
+#if _M_SSE >= 0x501
+/// On AVX2, uses the given broadcast to load from memory, otherwise uses the load
+# define BROADCAST_OR_LOAD(broadcast, load, dst, src) \
+	broadcast(dst, src)
+# define _rip_local_d(x) _rip_local(d8.x)
+# define _rip_local_d_p(x) _rip_local_d(p.x)
+#else
+/// On AVX2, uses the given broadcast to load from memory, otherwise uses a the load
```

Comment:
This comment is a bit confusing as the "else" isn't avx2. Maybe 1 comment before the if would be better. or 1 comment with avx2 broadcast and 2nd comment otherwise.

Reply:
The comment was meant to show up in IDEs if you moused over `BROADCAST_OR_LOAD`, no matter which mode your IDE thought you were compiling with at the given time

(So the comment describes `BROADCAST_OR_LOAD` itself, not the specific implementation)

Reply:
Oh ok. Not a clever IDE if comment need to be duplicated N times. It isn't important.

```
load128(reg, mem)

Reply:
Previously I had broadcasts of other sizes being used with BROADCAST_OR_LOAD but it seems they're all gone now so yeah I can switch it over

Reply:
Actually it I somehow missed DrawScanline needs the following pairings:
- vbroadcastf128/movaps
- vbroadcasti128/movdqa
- vbroadcastss/movaps
- vpbroadcastq/movdqa
- vpbroadcastd/movdqa
- vpbroadcastw/movdqa

I made the change to both this and the PR for DrawScanline, though we could do it for just one if you think that would be better